docs/proposals/20200615-external-remediation.md

```diff
@@ -0,0 +1,299 @@
+
+    kind: Metal3RemediationTemplate
+    apiVersion: remediation.metal3.io/v1alphaX
+    metadata:
+      name: M3_REMEDIATION_GROUP
+      namespace: NAMESPACE_OF_UNHEALTHY_MACHINE
+    spec:
+      template:
+        spec:
+          strategy: 		    escalate
+          deleteAfterRetries:     10
+          powerOnTimeoutSeconds:  600
+          powerOffTimeoutSeconds: 120
+
+Metal3Remediation:
+
+    apiVersion: remediation.metal3.io/v1alphaX
+    kind: Metal3Remediation
+    metadata:
+      name: NAME_OF_UNHEALTHY_MACHINE
+      namespace: NAMESPACE_OF_UNHEALTHY_MACHINE
+       finalizer: 
+       - remediation.metal3.io
+       ownerReferences: 
+    apiVersion:cluster.x-k8s.io/v1alphaX
+    kind: MachineHealthCheck
+    name: REMEDIATION_GROUP
+    uid: ...
+      spec: 
+        strategy: 		    escalate
+        deleteAfterRetries:     10
+        powerOnTimeoutSeconds:  600
+        powerOffTimeoutSeconds: 120
+      status:
+        phase: power off
+        retryCount: 1
+
+#### Usage
+
+    
+    func (t *healthCheckTarget) createProviderRemediation(ctx context.Context) (*corev1.ObjectReference, error) {
+        if t.MHC.Spec.RemediationTemplate == nil {
+            return nil, nil
+    }
+    
+    from, err := Get(ctx, c, t.MHC.Spec.RemediationTemplate, t.Machine.Namespace)
+    if err != nil {
+        return nil, err
+    }
+    generateTemplateInput := &GenerateTemplateInput{
+        Template:    from,
+        Namespace:   t.Machine.Namespace,
+        ClusterName: t.Machine.ClusterName,
+        OwnerRef:    metav1.OwnerReference{*metav1.NewControllerRef(
+    t.Machine, t.Machine.Kind)},
+        Labels:      [],
+    }
+    to, err := GenerateTemplate(generateTemplateInput)
+    if err != nil {
+        return nil, err
+    }
+    
+    // Force the same name as the machine
+    to.SetName(t.Machine.Name)
+    
+    // Create the external clone.
+    if err := r.Client.Create(context.Background(), to); err != nil {
+        return nil, err
+    }
+    return GetObjectReference(to), nil
+    }
+    
+    func (t *healthCheckTarget) remediate(...) error {
+        ...
+    remediation, err := t.createProviderRemediation(ctx)
+    if remediation != nil && err != nil {
+        // control passes to external remediation controller
+        return nil
+    } else if err != nil && errors.IsAlreadyExists(err) {
+        // external remediation controller already has control
+    return nil
+    } else if err != nil {
+        logger.Errorf("%q: failed to create provider remediation object for Machine %q in namespace %q", err, t.machine.Name, t.machine.Namespace)
+    }
+    ...
+    }
+
+### Risks and Mitigations
+
+- Currently an admin can configure multiple MHCs with overlapping selectors, meaning that he can configure different remediation templates to the same Machine. This will currently be resolved by proper documentation as described in [issue #3058](https://github.com/kubernetes-sigs/cluster-api/issues/3058)
```

Comment:
I would not consider this a risk that is created by or specific to this proposal. Want to remove it?